dotcom-rendering/src/lib/identity.ts

```diff
@@ -22,3 +28,159 @@ export async function isSignedInWithOktaAuthState(): Promise<
+	publicFields: {
+		displayName: string;
+	};
+	statusFields: {
+		userEmailValidated: boolean;
+	};
+	id: number;
+	rawResponse: string;
+} | null;
+
+let userFromCookieCache: IdentityUserFromCache = null;
+const cookieName = 'GU_U';
+const getUserCookie = (): string | null => getCookie({ name: cookieName });
+const decodeBase64 = (str: string): string =>
+	decodeURIComponent(
+		escape(
+			window.atob(
+				str.replace(/-/g, '+').replace(/_/g, '/').replace(/,/g, '='),
+			),
+		),
+	);
+
+const getUserFromCookie = (): IdentityUserFromCache => {
+	if (userFromCookieCache === null) {
+		const cookieData = getUserCookie();
+
+		if (cookieData) {
+			const userData = JSON.parse(
+				decodeBase64(cookieData.split('.')[0] ?? ''),
+			) as string[];
+
+			const id = parseInt(userData[0] ?? '', 10);
+			const displayName = decodeURIComponent(userData[2] ?? '');
+			const accountCreatedDate = userData[6];
+			const userEmailValidated = Boolean(userData[7]);
+
+			if (id && accountCreatedDate) {
+				userFromCookieCache = {
+					id,
+					publicFields: {
+						displayName,
+					},
+					dates: { accountCreatedDate },
+					statusFields: {
+						userEmailValidated,
+					},
+					rawResponse: cookieData,
+				};
+			}
+		}
+	}
+
+	return userFromCookieCache;
+};
+
+export const isUserLoggedIn = (): boolean => getUserFromCookie() !== null;
+
+export const getOptionsHeadersWithOkta = (
+	authStatus: SignedInWithCookies | SignedInWithOkta,
+): RequestInit => {
+	if (authStatus.kind === 'SignedInWithCookies') {
+		return {
+			credentials: 'include',
+		};
+	}
+
+	return {
+		headers: {
+			Authorization: `Bearer ${authStatus.accessToken.accessToken}`,
+			'X-GU-IS-OAUTH': 'true',
+		},
+	};
+};
+
+export async function getAuthState(): Promise<OktaAuthState> {
+	const authState = await isSignedInWithOktaAuthState();
+	return authState;
+}
+
+export async function eitherInOktaTestOrElse<A, B>(
+	inOktaTestFunction: (authState: OktaAuthState) => A,
+	notInOktaTestFunction: () => B,
+): Promise<A | B> {
+	const useOkta = !!window.guardian.config.switches.okta;
+
+	if (useOkta) {
+		const authState = await getAuthState();
+		return inOktaTestFunction(authState);
+	} else {
+		return notInOktaTestFunction();
+	}
+}
+
+export function getSignedInStatusWithOkta(
+	authState: OktaAuthState,
+): SignedOutWithOkta | SignedInWithOkta {
+	if (authState.isAuthenticated) {
+		return {
+			kind: 'SignedInWithOkta',
+			accessToken: authState.accessToken,
+			idToken: authState.idToken,
+		};
+	}
+
+	return { kind: 'SignedOutWithOkta' };
+}
+
+export const isUserLoggedInOktaRefactor = (): Promise<boolean> =>
+	getAuthStatus().then((authStatus) =>
+		authStatus.kind === 'SignedInWithCookies' ||
+		authStatus.kind === 'SignedInWithOkta'
+			? true
+			: false,
+	);
+
+export function getSignedInStatusWithCookies():
+	| SignedOutWithCookies
+	| SignedInWithCookies {
+	const GU_UCookie = getCookie({ name: 'GU_U', shouldMemoize: true });
```

Comment:
Not a blocker: strange that we have `getUserCookie` and not use it here.

We should probably unify these calls.

Reply:
looks like a nice refinement, Ill address and test 👍

Reply:
We should probably understand why there’s a difference and whether we actually want to memoize or not…

Reply:
removed `getUserCookie` as unnecessary, I think due the refresh function being invoked several times, caching the result of getCookie for the same name and not recomputing it every time, can avoid redundant computations

Reply:
I actually think that having two different ways of reading the cookie is more error-prone: you now have to change the name of the cookie `GU_U` in two places instead of one in this very file.

Reply:
great have centralised this, taking advantage of memoization throughout 👍